text/0232-maturity-of-otel.md

```diff
@@ -0,0 +1,72 @@
+
+Components should not be marked as stable if their user-visible interfaces are not stable. For instance, if the Collector's component "otlpreceiver" declares a dependency on the OpenTelemetry Collector API "config" package which is marked with a maturity level of "beta", the "otlpreceiver" should be at most "beta". Maintainers are free to deviate from this recommendation if they believe users are not going to be affected by future changes.
+
+For the purposes of this document, a breaking change is defined as a change that may require consumers of our components to adapt themselves in order to avoid disruption to their usage of our components.
+
+#### Development
+
+Not all pieces of the component are in place yet, and it might not be available for users yet. Bugs and performance issues are expected to be reported. User feedback is desired, especially regarding the user experience (configuration options, component observability, technical implementation details, and so on). Configuration options might break often depending on how things evolve. The component SHOULD NOT be used in production. The component MAY be removed without prior notice.
+
+#### Alpha
+
+This is the default level: any components with no explicit maturity level should be assumed to be "Alpha". The component is ready to be used for limited non-critical production workloads, and the authors of this component welcome user feedback. Bugs and performance problems are encouraged to be reported, but component owners might not work on them immediately. The component's interface and configuration options might often change without backward compatibility guarantees. Components at this stage might be dropped at any time without notice.
+
+#### Beta
+
+Same as Alpha, but the interfaces (API, configuration, generated telemetry) are treated as stable whenever possible. While there might be breaking changes between releases, component owners should try to minimize them. A component at this stage is expected to have had exposure to non-critical production workloads already during its Alpha phase, making it suitable for broader usage.
+
+#### Release Candidate
+
+The component is feature-complete and ready for broader usage. The component is ready to be declared stable, it might just need to be tested in more production environments before that can happen. Bugs and performance problems are expected to be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
+
+#### Stable
+
+The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
```

Comment:
once stable, breaking changes should not be occurring in any circumstances, with or without prior notice, right?

Reply:
No, that's not what the text says. The text says:
> Breaking changes, including configuration options and the component's output, are not expected to happen without prior notice unless under special circumstances.
Decomposing this:
> Breaking changes are not expected to happen without prior notice.
Meaning, we will still have breaking changes, but we will organize ourselves around them so they have ample notice. Unless special circumstances which are not defined in this doc.

Reply:
Maybe this sentence should avoid using negatives. A rephrasing might make sense.
```suggestion
The component is ready for general availability. Bugs and performance problems should be reported, and there's an expectation that the component owners will work on them. Changes to configuration options and the component's output are considered breaking changes. Breaking changes are expected to happen with prior notice unless under special circumstances.
```

Reply:
or maybe say - Breaking can occur, as long as semver guidelines are followed? (i.e bump major version.)

Reply:
None of the guidelines here talk about semantic versioning. I guess it's a mix of NIH, trying to steer away from versioning discussions directly, and giving components some flexibility. Add to that Go versioning policies are really strict and will enforce semver in ways that might not feel welcome. You could definitely offer notice by applying the semantic versioning by associating a breaking change with a major version change.

Reply:
> Breaking changes are expected to happen with prior notice unless under special circumstances.

This flips the meaning of the original text. Breaking changes should be very rare, we should always strive to not have them at all, but we acknowledge that special circumstances might warrant it. The text explicitly attempts to stay away from what's a special circumstance but think of it like a security problem behind a default value.

Reply:
OK, that’s a bit concerning. We need to clarify this.

Reply:
I see the confusion now:

> without prior notice unless under special circumstances

When taking this as the context, it gives the understanding that prior notice might not be needed under special circumstances. This is wrong. The "unless" refers to the whole statement:

> are not expected to happen without prior notice unless under special circumstances

I tweaked the wording a bit here in an attempt to make it clearer. Can you check again?

Reply:
yes, it's clearer. I was really confused by that sentence. I find it hard to believe we're going to say no more breaking changes once stable though. How do you evolve software without breaking changes (unless special circumstances?) ? Is explicit versioning something we can address maybe down the road to allow folks to give themselves some breathing room?

Reply:
This OTEP is here to help establish the maturity for deliverables within SIG, and I would argue that a Collector 1.0.0 is one deliverable, while 2.0RC1 is another. They can have different stability levels, while still being part of the same SIG.

Reply:
@atoulme, @cijothomas, is there something to clarify around this, or can this be marked as resolved?

Reply:
So versioning is key. Thanks Juraci, we're good here then.